numbagg/decorators.py

```diff
@@ -281,11 +283,23 @@ def _create_gufunc(self, core_ndim):
         return vectorize(self.func)
 
     def __call__(self, values, labels, axis=None, num_labels=None):
+
         values = np.asarray(values)
         labels = np.asarray(labels)
 
+        # passing num_lables avoids checking for factorized labels
         if num_labels is None:
-            num_labels = np.max(labels) + 1
+            if is_factorized(labels):
```

Comment:
Just to be clear, this only an optimization for the case where `labels` are already the integers [0, 1, 2,  ..., n-1]? Let's be sure to add a comment to that effect (but to be honest, I'm not sure it's worth it).

There also seems to be a difference here with regard to the handling of multi-dimensional inputs.

Reply:
Rather than using `num_labels is None` and then inspecting the array to decide if factorization is needed, what about adding an optional argument `factorized=False`? Or even better, maybe having a completely separate method for processing pre-factorized labels?

Then we wouldn't need to inspect the inputs for factorization under any circumstances.

Reply:
I agree re the separate method. I should have responded earlier to your comment. I think we could revert to the old method (which requires the input to be factorized), and offer helper methods to factorize data

Reply:
How about:
`result, uniques = numbagg.group_nanmean(values, labels, axis)`
and
`result = numbagg.factorized_group_nanmean(values, codes, axis, num_codes)`
?

There are various ways to imagine implementing this. Probably the simplest would be to make a method `compute_factorized` (used internally by `__call__`) and then set `factorized_group_nanmean = groupby_nanmean.compute_factorized` in grouped.py.

Reply:
Another option:
`result = numbagg.group_nanmean(values, codes, axis, num_codes)`
and 
`result, unique = numbagg.factorizing(numbagg.group_nanmean)(values, labels, axis)`

Reply:
> `result, unique = numbagg.factorizing(numbagg.group_nanmean)(values, labels, axis)`
> `result, unique = numbagg.group_nanmean.factorizing(values, labels, axis)`

...those are great ideas!

Consistent with those, I do think it could make sense to keep the factorization as part of the grouping function / some modified grouping function, given that factorizing requires both a factorize step and an 'unfactorize' step - the user is expecting their original labels back, not the ints that `group_nanmean` use